### PR TITLE
8269034: AccessControlException for SunPKCS11 daemon threads

### PR DESCRIPTION
**Pull request by coffeys**
Sufficient permissions missing if this code was ever to run with SecurityManager. 

Cleanest approach appears to be use of InnocuousThread to create the cleaner/poller threads.
Test case coverage extended to cover the SecurityManager scenario.

Reviewer request: @valeriepeng

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269034](https://bugs.openjdk.java.net/browse/JDK-8269034): AccessControlException for SunPKCS11 daemon threads

### Reviewers
 * [Valerie Peng](https://openjdk.java.net/census#valeriep) (@valeriepeng - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/117/head:pull/117` \
`$ git checkout pull/117`

Update a local copy of the PR: \
`$ git checkout pull/117` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/117/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 117`

View PR using the GUI difftool: \
`$ git pr show -t 117`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/117.diff">https://git.openjdk.java.net/jdk17/pull/117.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/module-info.java [hunk 1]
```diff
@@ -199,6 +199,7 @@
         jdk.attach,
         jdk.charsets,
         jdk.compiler,
+        jdk.crypto.cryptoki,
```
Comment (AlanBateman) [inline src/java.base/share/classes/module-info.java:202]: At some point I'd like to see this re-visited so we can avoid exporting this package to modules defined to the platform class loader.

File: test/jdk/sun/security/pkcs11/Provider/MultipleLogins.java [hunk 2]
```diff
@@ -117,6 +123,27 @@ private static void test(SunPKCS11 p) throws Exception {
...
+        final Permissions perms = new Permissions();
+        SimplePolicy() {
+            perms.add(new PropertyPermission("*", "read, write"));
+            perms.add(new RuntimePermission("accessClassInPackage.sun.*"));
+            perms.add(new RuntimePermission("accessClassInPackage.sun.security.pkcs11.*"));
+            perms.add(new SecurityPermission("authProvider.*"));
+            perms.add(new SecurityPermission("clearProviderProperties.*"));
+            perms.add(new SecurityPermission("insertProvider.*"));
+            perms.add(new SecurityPermission("removeProvider.*"));
+        }
```
Comment (valeriepeng) [inline test/jdk/sun/security/pkcs11/Provider/MultipleLogins.java:137]: The test still pass without the following permission:
```
             perms.add(new RuntimePermission("accessClassInPackage.sun.*"));
             perms.add(new RuntimePermission("accessClassInPackage.sun.security.pkcs11.*"));
             perms.add(new SecurityPermission("clearProviderProperties.*"));
```
Remove them?

File: test/jdk/sun/security/pkcs11/Provider/MultipleLogins.sh [hunk 3]
```diff
@@ -126,10 +126,21 @@ ${TESTJAVA}${FS}bin${FS}java ${TESTVMOPTS} \
...
+        --add-modules jdk.crypto.cryptoki \
+        --add-exports jdk.crypto.cryptoki/sun.security.pkcs11=ALL-UNNAMED \
+        -DCUSTOM_DB_DIR=${TESTCLASSES} \
+        -DCUSTOM_P11_CONFIG=${TESTSRC}${FS}MultipleLogins-nss.txt \
+        -DNO_DEFAULT=true \
+        -DNO_DEIMOS=true \
+        -Dtest.src=${TESTSRC} \
+        -Dtest.classes=${TESTCLASSES} \
+        -Djava.security.debug=${DEBUG} \
+        MultipleLogins ${TESTSRC}${FS}MultipleLogins.policy || exit 11
```
Comment (valeriepeng) [inline test/jdk/sun/security/pkcs11/Provider/MultipleLogins.sh:143]: There is no MultipleLogins.policy file. The test just uses the internal SimplePolicy object. Maybe just use a string like "useSimplePolicy".

File: test/jdk/sun/security/pkcs11/Provider/MultipleLogins.sh [hunk 4]
```diff
@@ -126,10 +126,21 @@ ${TESTJAVA}${FS}bin${FS}java ${TESTVMOPTS} \
...
+        -classpath ${TESTCLASSPATH} \
+        --add-modules jdk.crypto.cryptoki \
+        --add-exports jdk.crypto.cryptoki/sun.security.pkcs11=ALL-UNNAMED \
+        -DCUSTOM_DB_DIR=${TESTCLASSES} \
+        -DCUSTOM_P11_CONFIG=${TESTSRC}${FS}MultipleLogins-nss.txt \
+        -DNO_DEFAULT=true \
+        -DNO_DEIMOS=true \
+        -Dtest.src=${TESTSRC} \
+        -Dtest.classes=${TESTCLASSES} \
+        -Djava.security.debug=${DEBUG} \
```
Comment (valeriepeng) [inline test/jdk/sun/security/pkcs11/Provider/MultipleLogins.sh:142]: Save these java options and use it for both invocation? This way it's easier to tell that there is no difference among these two except for the extra argument.

File: test/jdk/sun/security/pkcs11/Provider/MultipleLogins.java [hunk 5]
```diff
@@ -44,19 +43,27 @@
...
-            providers[i] =
-                    (SunPKCS11)PKCS11Test.newPKCS11Provider()
-                    .configure(nssConfig);
+            providers[i] = (SunPKCS11)PKCS11Test.newPKCS11Provider();
+        }
+
+        if (args.length > 0) {
+            Policy.setPolicy(new SimplePolicy());
+            System.setSecurityManager(new SecurityManager());
+        }
```
Comment (valeriepeng) [inline test/jdk/sun/security/pkcs11/Provider/MultipleLogins.java:63]: Just curious, why split the loop into 2 and set the SecurityManager in between the two loops? Can't we just set the policy/security manager before the loop?

**Review by valeriepeng — APPROVED** [review]
Update looks good. Thanks, Valerie

**Pull request closed without merging**